### DISCUSSION TITLE
Vue: how to spy on Component method?

### DISCUSSION THREAD

**Discussion by nathan-kerr** [cypress-io/cypress#30367] [Questions and Help]
My spy is never called even though I can visually confirm that the target method `getRandomTerm()` is called when mounted. The image below shows that a term was generated and displayed in `<h1>`.

VocabFlashCard.vue:
```vue
<template>
    <h1>{{ currentTerm.name }}</h1>
</template>

<script setup lang="ts">
const currentTerm = ref<Term>({ name: '', image: '' })

function getRandomTerm() {
  currentTerm.value = terms.terms[Math.floor(Math.random() * terms.terms.length)]
}

onMounted(() => {
  getRandomTerm()
})
</script>
```

VocabFlashCard.cy.ts:
```ts

beforeEach(() => {
  const getRandomTermSpy = cy.spy().as('getRandomTermSpy')

  mount(VocabFlashCard, {
    props: {
      terms: testTerms,    
      getRandomTerm: getRandomTermSpy
    }
  })
})

it('generates a random term', () => {
  cy.get('@getRandomTermSpy').should('have.been.calledOnce')
})
```

![image](https://github.com/user-attachments/assets/98f29fa5-8fe9-48e1-a586-a328f5cc0b8f)

**Comment by portyu9** [comment]
Hi @nathan-kerr 

The spy is never called because the component never calls the function you passed as the `getRandomTerm` prop.

Inside `<script setup>`, this:

```ts
function getRandomTerm() {
  currentTerm.value =
    terms.terms[Math.floor(Math.random() * terms.terms.length)]
}

onMounted(() => {
  getRandomTerm()
})
```

calls the **local lexical function** `getRandomTerm()`.

This in the test:

```ts
props: {
  terms: testTerms,
  getRandomTerm: getRandomTermSpy
}
```

only passes another value as a prop. Unless the component declares and explicitly calls that prop, it cannot intercept the local function call.

For this component, I would test the observable behavior instead:

```ts
mount(VocabFlashCard, {
  props: {
    terms: testTerms,
  },
})

cy.get('h1')
  .invoke('text')
  .should((text) => {
    expect(testTerms.map((term) => term.name)).to.include(text)
  })
```

If you specifically need call-level testing, extract the random-term selection into a separate function/composable and unit-test it, or explicitly inject a function dependency that the component actually calls.

`defineExpose()` by itself would not solve this: exposing the method does not change the local `getRandomTerm()` reference that `onMounted()` calls.

I could not retrieve the attached screenshot because its GitHub attachment URL now returns 404, so I did not rely on the screenshot for this diagnosis.

Please mark this answer as accepted if it helped, thank you.
